File: apps/desktop/src-tauri/src/downloads/subs.rs
//! Sous-titres texte en side-cars : téléchargés via le proxy (conversion de
//! format assurée par Jellyfin — srt/ass/vtt), enregistrés sous
//! `media/<itemId>/subs/`. Les sous-titres image (PGS/VobSub) ne sont PAS
//! convertibles en texte : hors ligne, ils n'existent qu'incrustés (burn-in)
//! dans une variante Allégée. Best-effort : un sous-titre manquant ne bloque
//! jamais le média.

use super::fsops;
use serde::{Deserialize, Serialize};
use std::io::Read;
use std::path::Path;

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleSpec {
    /// Index du MediaStream côté Jellyfin.
    pub index: i64,
    /// Format cible : srt | ass | vtt (validé ici).
    pub format: String,
    /// Étiquette de fichier pré-construite côté client (ex. "fre-forced") —
    /// re-sanitisée par prudence.
    pub lang_tag: String,
}

fn sanitize_tag(tag: &str) -> String {
    let cleaned: String = tag
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == '-')
        .take(40)
        .collect();
    if cleaned.is_empty() { "und".into() } else { cleaned.to_ascii_lowercase() }
}

fn valid_format(format: &str) -> bool {
    matches!(format, "srt" | "ass" | "vtt")
}

pub fn parse_specs(json: &str) -> Vec<SubtitleSpec> {
    serde_json::from_str::<Vec<SubtitleSpec>>(json).unwrap_or_default()
}

/// Télécharge chaque sous-titre texte. Retourne le nombre récupéré.
pub fn fetch_all(
    agent: &ureq::Agent,
    server_url: &str,
    token: &str,
    root: &Path,
    item_id: &str,
    media_source_id: &str,
    specs: &[SubtitleSpec],
) -> usize {
    let mut fetched = 0;
    for spec in specs {
        if !valid_format(&spec.format) || spec.index < 0 {
            continue;
        }
        let rel = format!(
            "media/{item_id}/subs/{}-{}.{}",
            spec.index,
            sanitize_tag(&spec.lang_tag),
            spec.format
        );
        let Ok(path) = fsops::safe_join(root, &rel) else { continue };
        if path.exists() {
            fetched += 1;
            continue;
        }
        let url = format!(
            "{server_url}/api/jellyfin/Videos/{item_id}/{media_source_id}/Subtitles/{}/Stream.{}",
            spec.index, spec.format
        );
        let Ok(response) = agent
            .get(&url)
            .set("Authorization", &format!("Bearer {token}"))
            .call()
        else {
            continue;
        };
        let mut bytes = Vec::new();
        if response
            .into_reader()
            .take(15 * 1024 * 1024)
            .read_to_end(&mut bytes)
            .is_err()
            || bytes.is_empty()
        {
            continue;
        }
        if let Some(parent) = path.parent() {
            if std::fs::create_dir_all(parent).is_err() {
                continue;
            }
        }
        if std::fs::write(&path, &bytes).is_ok() {
            fetched += 1;
        }
    }
    fetched
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitisation_des_etiquettes() {
        assert_eq!(sanitize_tag("fre-forced"), "fre-forced");
        assert_eq!(sanitize_tag("../evil/FR"), "evilfr");
        assert_eq!(sanitize_tag(""), "und");
        assert!(sanitize_tag(&"x".repeat(100)).len() <= 40);
    }

    #[test]
    fn parse_specs_tolere_le_json_invalide() {
        assert!(parse_specs("pas du json").is_empty());
        let specs = parse_specs(r#"[{"index":3,"format":"srt","langTag":"fre"}]"#);
        assert_eq!(specs.len(), 1);
        assert_eq!(specs[0].index, 3);
        assert_eq!(specs[0].format, "srt");
    }
}

File: apps/desktop-electron/src/main/downloads/core/subs.ts
/**
 * Sous-titres texte en side-cars.
 *
 * Téléchargés par le proxy — c'est Jellyfin qui convertit le format — et
 * enregistrés sous `media/<itemId>/subs/`. Les sous-titres IMAGE (PGS, VobSub)
 * ne sont pas convertibles en texte : hors ligne, ils n'existent qu'incrustés
 * dans une variante Allégée.
 *
 * Best-effort de bout en bout : un sous-titre manquant ne bloque jamais le
 * média.
 *
 * Portage de `apps/desktop/src-tauri/src/downloads/subs.rs`.
 */

import type { Volume } from "./adapters";
import { MAX_SUBTITLE_BYTES, type FetchBytes } from "./fetcher";
import { asInteger, asString, field, parseJsonText } from "./json";
import { safeJoin } from "./paths";

export interface SubtitleSpec {
  /** Index du `MediaStream` côté Jellyfin. */
  index: number;
  /** Format cible : `srt`, `ass` ou `vtt`. */
  format: string;
  /** Étiquette de fichier construite côté client (« fre-forced »). */
  langTag: string;
}

const FORMATS = new Set(["srt", "ass", "vtt"]);

/**
 * L'étiquette entre dans un NOM DE FICHIER : elle est re-nettoyée ici même si
 * le client l'a déjà fait. Un identifiant qui ne laisse rien devient `und`
 * plutôt qu'une chaîne vide, qui collisionnerait entre pistes.
 */
export function sanitizeTag(tag: string): string {
  const clean = [...tag]
    .filter((c) => /[A-Za-z0-9-]/.test(c))
    .slice(0, 40)
    .join("");
  return clean === "" ? "und" : clean.toLowerCase();
}

/** Analyse la liste stockée en base. Tolère un JSON abîmé. */
export function parseSpecs(json: string): SubtitleSpec[] {
  const raw = parseJsonText(json);
  if (!Array.isArray(raw)) return [];
  const specs: SubtitleSpec[] = [];
  for (const entry of raw) {
    const index = asInteger(field(entry, "index"));
    const format = asString(field(entry, "format"));
    const langTag = asString(field(entry, "langTag"));
    if (index === null || format === null) continue;
    specs.push({ index, format, langTag: langTag ?? "und" });
  }
  return specs;
}

/** Chemin relatif d'un side-car. */
export function subtitleRelPath(itemId: string, spec: SubtitleSpec): string {
  return `media/${itemId}/subs/${spec.index}-${sanitizeTag(spec.langTag)}.${spec.format}`;
}

/** Pourquoi une piste n'est pas arrivée — assez pour distinguer le serveur du disque. */
export type SubtitleSkip = "format" | "path" | "fetch" | "write";

/**
 * Télécharge chaque sous-titre texte manquant. Retourne le nombre disponible
 * en fin d'opération — un fichier déjà là compte, la fonction est idempotente
 * et c'est ce qui permet à la réparation de la rappeler sans coût.
 *
 * `onSkipped` est la seule trace d'une piste perdue. Sur une série dont le
 * serveur extrait chaque sous-titre à coups de ffmpeg, une extraction lente
 * dépasse le délai du fetcher et rendait `null` — sauté sans un mot.
 */
export async function fetchAll(
  fetchBytes: FetchBytes,
  serverUrl: string,
  volume: Volume,
  itemId: string,
  mediaSourceId: string,
  specs: readonly SubtitleSpec[],
  onSkipped?: (spec: SubtitleSpec, reason: SubtitleSkip) => void,
): Promise<number> {
  let fetched = 0;
  for (const spec of specs) {
    if (!FORMATS.has(spec.format) || spec.index < 0) {
      onSkipped?.(spec, "format");
      continue;
    }

    let target: string;
    try {
      target = safeJoin(volume, subtitleRelPath(itemId, spec));
    } catch {
      onSkipped?.(spec, "path");
      continue;
    }
    if (volume.files.exists(target)) {
      fetched += 1;
      continue;
    }

    const url =
      `${serverUrl}/api/jellyfin/Videos/${itemId}/${mediaSourceId}` +
      `/Subtitles/${spec.index}/Stream.${spec.format}`;
    const bytes = await fetchBytes(url, MAX_SUBTITLE_BYTES);
    if (bytes === null || bytes.byteLength === 0) {
      // Délai dépassé, refus du serveur, corps vide : le fetcher les confond
      // tous en `null`, et la réparation repassera au retour du réseau.
      onSkipped?.(spec, "fetch");
      continue;
    }

    try {
      volume.files.mkdirp(volume.files.dirname(target));
      volume.files.writeBytes(target, bytes);
      fetched += 1;
    } catch {
      // Disque plein ou droits : le média reste lisible sans ses sous-titres.
      onSkipped?.(spec, "write");
    }
  }
  return fetched;
}
